[SRVKS-869] Document gRPC over HTTP2 for 4.10+ (#58177)
openshift/openshift-docs: modules/interacting-serverless-apps-http2-gRPC-up-to-4-9.adoc

----
+
.Example output
+
[source,terminal]
----
NAME                   TYPE           CLUSTER-IP      EXTERNAL-IP                                                             PORT(S)                                                                                                                                      AGE
kourier   LoadBalancer   172.30.51.103   a83e86291bcdd11e993af02b7a65e514-33544245.us-east-1.elb.amazonaws.com   80:31380/TCP,443:31390/TCP   67m
----
+
The public address is surfaced in the `EXTERNAL-IP` field, and in this case is `a83e86291bcdd11e993af02b7a65e514-33544245.us-east-1.elb.amazonaws.com`.

. Manually set the host header of your HTTP request to the application's host, but direct the request itself against the public address of the ingress gateway.
+
[source,terminal]
----
$ curl -H "Host: hello-default.example.com" a83e86291bcdd11e993af02b7a65e514-33544245.us-east-1.elb.amazonaws.com
----
+
.Example output
[source,terminal]
----
Hello Serverless!
----
+
You can also make a gRPC request by setting the authority to the application's host, while directing the request against the ingress gateway directly:
+
[source,yaml]

> ----

Reviewer: Hmm... Actually here is messing up. (Just for excuse, but we didn't add this section but just moving the location :sweat_smile: )

I guess this section wants to explain that:
* We can set the LoadBalancer type to the gateway service.
* The External-IP can be accessed directory via HTTP.
* The External-IP can be accessed directory via gRPC (if backend has gRPC app) because it does not pass through OCP Route.

> ----

Reviewer 2: OK. I think I get it. So, I'd change the sentence above from:
```
You can also make a gRPC request by setting the authority to the application's host, while directing the request against the ingress gateway directly
```
to
```
You can also make a direct gRPC request against the ingress gateway.
```
Let's leave out the "setting the authority" part from that sentence because that's just confusing and not important here.
And let's also include the import so that users know which library we're using.

Makes sense?

> ----

Reviewer: ```suggestion
----
import "google.golang.org/grpc"
```

Let's add `import "google.golang.org/grpc"` as Martin suggested. We can fix this section or at least drop this gRPC workaround after 1.29 release (as 4.9- are not supported).

> ----

Reviewer: Sounds good to me :+1:

// Module included in the following assemblies:
//
// serverless/knative-serving/external-ingress-routing/using-http2-gRPC.adoc

:_content-type: PROCEDURE

> [id="interacting-serverless-apps-http2-gRPC-up-to-4-9_{context}"]

Reviewer 3: IDs should not have capital letters in them (my understanding is that they can break Pantheon builds) . Can you make these lowercase?

> = Interacting with a serverless application using HTTP2 and gRPC in OpenShift Dedicated 4.9 and older

Reviewer 4: ```suggestion
= Interacting with a serverless application using HTTP2 and gRPC in {product-title} Dedicated 4.9 and older
```

Author: @skrthomas applied your suggestion as such:
`in {product-title} 4.9`  (results into in _OpenShift Dedicated 4.9_)
if we don't omit `Dedicated`, we get it twice:
`in {product-title} Dedicated 4.9`  (results into in _OpenShift Dedicated Dedicated 4.9_)

[IMPORTANT]
====
This method needs to expose Kourier Gateway using the `LoadBalancer` service type. You can configure this by adding the following YAML to your `KnativeServing` custom resource definition (CRD):

[source,yaml]
----
...
spec:
  ingress:
    kourier:
      service-type: LoadBalancer
...
----
====

.Prerequisites

* {ServerlessOperatorName} and Knative Serving are installed on your cluster.
* Install the OpenShift CLI (`oc`).

> * You have created a Knative service.

Reviewer 3: Normally I wouldn't point this out, but since you have to fix the xref above, would you mind mind making these tenses more consistent in your prerequisite sections? There are currently three tenses being use: 

* Present tense 
* Imperative mood 
* Present perfect 

I think maybe using present perfect is the best, eg 

* YOu have installed {ServerlessOperatorName} and Knative Serving d on your cluster.
* You have installed the OpenShift CLI
* You have created a Knative service.

. Find the ingress gateway's public address:
+
[source,terminal]
----
$ oc -n knative-serving-ingress get svc kourier
----
+
.Example output
+
[source,terminal]
----
NAME                   TYPE           CLUSTER-IP      EXTERNAL-IP                                                             PORT(S)                                                                                                                                      AGE
kourier   LoadBalancer   172.30.51.103   a83e86291bcdd11e993af02b7a65e514-33544245.us-east-1.elb.amazonaws.com   80:31380/TCP,443:31390/TCP   67m
----
+
The public address is surfaced in the `EXTERNAL-IP` field, and in this case is `a83e86291bcdd11e993af02b7a65e514-33544245.us-east-1.elb.amazonaws.com`.

. Manually set the host header of your HTTP request to the application's host, but direct the request itself against the public address of the ingress gateway.
+
[source,terminal]
----
$ curl -H "Host: hello-default.example.com" a83e86291bcdd11e993af02b7a65e514-33544245.us-east-1.elb.amazonaws.com
----
+
.Example output
[source,terminal]
----
Hello Serverless!
----
+
You can also make a direct gRPC request against the ingress gateway:
+

> [source,yaml]

Reviewer 2: Is there an option to specify Golang here instead of yaml?

Reviewer: Ugh... Sorry I suggested this. We should use golang as:

```suggestion
[source,golang]
```

(It was used in other places so it should work.)

Author: Thanks, good catch! Fixed in both places.
